File: Chapter10/Java/MinHeap.java
package main.java;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by William Smith on 10/2/16.
 */
public class MinHeap
{
    List<HeapNode> elements;

    public int size()
    {
        return elements.size();
    }

    public MinHeap()
    {
        elements = new ArrayList<HeapNode>();
    }

    public void insert(HeapNode item)
    {
        elements.add(item);
        orderHeap();
    }

    public void delete(HeapNode item)
    {
        int i = elements.indexOf(item);
        int last = elements.size() - 1;

        elements.set(i, elements.get(last));
        elements.remove(last);
        orderHeap();
    }

    public HeapNode extractMin() //Pop
    {
        if (elements.size() > 0)
        {
            HeapNode item = elements.get(0);
            delete(item);
            return item;
        }

        return null;
    }

    public HeapNode findMin() //Peek
    {
        if (elements.size() > 0)
        {
            return elements.get(0);
        }

        return null;
    }

    private void orderHeap()
    {
        for (int i = elements.size() - 1; i > 0; i--)
        {
            int parentPosition = (i - 1) / 2;

            if (elements.get(parentPosition).Data > elements.get(i).Data)
            {
                swapElements(parentPosition, i);
            }
        }
    }

    private void swapElements(int firstIndex, int secondIndex)
    {
        HeapNode tmp = elements.get(firstIndex);
        elements.set(firstIndex, elements.get(secondIndex));
        elements.set(secondIndex, tmp);
    }

    public List<HeapNode> getChildren(int parentIndex)
    {
        if (parentIndex >= 0)
        {
            ArrayList<HeapNode> children = new ArrayList<HeapNode>();
            int childIndexOne = (2 * parentIndex) + 1;
            int childIndexTwo = (2 * parentIndex) + 2;
            children.add(elements.get(childIndexOne));
            children.add(elements.get(childIndexTwo));

            return children;
        }

        return null;
    }

    public HeapNode getParent(int childIndex)
    {
        if (childIndex > 0 && elements.size() > childIndex)
        {
            int parentIndex = (childIndex - 1) / 2;
            return elements.get(parentIndex);
        }

        return null;
    }

}

File: Chapter10/Swift/MinHeap.swift
//
//  MinHeap.swift
//  EverydayDataStructures-Swift
//
//  Created by William Smith on 10/6/16.
//  Copyright © 2016 Websmiths, LLC. All rights reserved.
//

import Foundation

public class MinHeap
{
    public var _elements: Array = [HeapNode]()


    public init () {}

    public func getCount() -> Int
    {
        return _elements.count
    }

    public func insert(item: HeapNode)
    {
        _elements.append(item)
        orderHeap()
    }

    public func delete(item: HeapNode)
    {
        if let index = _elements.index(of: item)
        {
            _elements[index] = _elements.last!
            _elements.removeLast()
            orderHeap()
        }
    }

    public func extractMin() -> HeapNode?
    {
        if (_elements.count > 0)
        {
            let item = _elements[0]
            delete(item: item)
            return item
        }
        
        return nil
    }

    public func findMin() -> HeapNode?
    {
        if (_elements.count > 0)
        {
            return _elements[0]
        }
        
        return nil
    }

    public func orderHeap()
    {
        for i in (0..<(_elements.count) - 1).reversed()
        {
            let parentPosition = (i - 1) / 2

            if (_elements[parentPosition].data! > _elements[i].data!)
            {
                swapElements(first: parentPosition, second: i)
            }
        }
    }

    public func swapElements(first: Int, second: Int)
    {
        let tmp = _elements[first]
        _elements[first] = _elements[second]
        _elements[second] = tmp
    }

    public func getChildren(parentIndex: Int) -> [HeapNode]?
    {
        if (parentIndex >= 0)
        {
            var children: Array = [HeapNode]()
            
            let childIndexOne = (2 * parentIndex) + 1;
            let childIndexTwo = (2 * parentIndex) + 2;
            children.append(_elements[childIndexOne])
            children.append(_elements[childIndexTwo])
            
            return children;
        }
        
        return nil;
    }

    public func getParent(childIndex: Int) -> HeapNode?
    {
        if (childIndex > 0 && _elements.count > childIndex)
        {
            let parentIndex = (childIndex - 1) / 2;
            return _elements[parentIndex];
        }
        
        return nil;
    }
}

public class HeapNode : Equatable
{
    public var data: Int?
    
    public init() {}
}

public func == (lhs: HeapNode, rhs: HeapNode) -> Bool {
    return (lhs.data == rhs.data)
}
